ledger/accountdb.go

```diff
@@ -4826,6 +4681,156 @@ func loadTxTail(ctx context.Context, tx *sql.Tx, dbRound basics.Round) (roundDat
 	return roundData, roundHash, expectedRound + 1, nil
 }
 
+func deleteStoredCatchpoint(ctx context.Context, e db.Executable, round basics.Round) error {
+	f := func() error {
+		query := "DELETE FROM storedcatchpoints WHERE round=?"
+		_, err := e.ExecContext(ctx, query, round)
+		return err
+	}
+	return db.Retry(f)
+}
+
+func storeCatchpoint(ctx context.Context, e db.Executable, round basics.Round, fileName string, catchpoint string, fileSize int64) error {
+	f := func() error {
+		query := "DELETE FROM storedcatchpoints WHERE round=?"
+		_, err := e.ExecContext(ctx, query, round)
+		if err != nil || (fileName == "" && catchpoint == "" && fileSize == 0) {
+			return err
+		}
+
+		query = "INSERT INTO storedcatchpoints(round, filename, catchpoint, filesize, pinned) VALUES(?, ?, ?, ?, 0)"
+		_, err = e.ExecContext(ctx, query, round, fileName, catchpoint, fileSize)
+		return err
+	}
+	return db.Retry(f)
+}
+
+func getOldestCatchpointFiles(ctx context.Context, q db.Queryable, fileCount int, filesToKeep int) (map[basics.Round]string /*fileNames*/, error) {
+	var res map[basics.Round]string
+
+	f := func() error {
+		query := "SELECT round, filename FROM storedcatchpoints WHERE pinned = 0 and round <= COALESCE((SELECT round FROM storedcatchpoints WHERE pinned = 0 ORDER BY round DESC LIMIT ?, 1),0) ORDER BY round ASC LIMIT ?"
+		rows, err := q.QueryContext(ctx, query, filesToKeep, fileCount)
+		if err != nil {
+			return err
+		}
+		defer rows.Close()
+
+		res = make(map[basics.Round]string)
+		for rows.Next() {
+			var fileName string
+			var round basics.Round
+			err = rows.Scan(&round, &fileName)
+			if err != nil {
+				return err
+			}
+			res[round] = fileName
+		}
+
+		return rows.Err()
+	}
+	err := db.Retry(f)
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func readCatchpointStateUint64(ctx context.Context, q db.Queryable, stateName catchpointState) (uint64, error) {
```

Comment:
as before, I'd prefer to have the old code with a single query change (prepared vs ad-hoc).